### ISSUE TITLE
Deserialize sealed classes generated by oneof

### ISSUE THREAD

**Issue by anjanbk** [mscheong01/krotoDC#24]
For example, suppose we have:

```
message TestWrapper {
  oneof data {
    string typeA = 1;
    string typeB = 2;
  }
}
```

Which generates:

```
@KrotoDC(forProto = TestWrapper::class)
public data class TestWrapper(
  public val `data`: Data? = null,
) {
  public sealed interface Data {
    public data class TypeA(
      public val typeA: String = "",
    ) : Data

    public data class TypeB(
      public val typeB: String = "",
    ) : Data
  }
}
```

By default, this serializes to:

```
"wrapper": {
  "data": {
    "typeA": "foobar"
  }
},
```

However it cannot be deserialized by Gson without specifying an InstanceCreator or TypeAdapter for that specific type.

Is there a way to handle serialization/deserialization of oneofs that I'm missing, or is this just something that's not yet supported by the KrotoDC code generation?

**Comment by mscheong01** [comment]
Since krotoDC generated classes can be converted to protobuf-java classes, you can deserialize the json to protobuf-java class and simply call .toDataClass() (extension function generated by krotoDC).
If this does not work for your usecase, please let me know 😄

**Comment by mscheong01** [comment]
The serialized Protobuf JSON you provided seems to be a direct JSON serialization of the Protobuf Java class, leading to the exposure of hidden fields such as `memoizedHashCode`, etc., which I believe is the root of your problem.

Have you considered using the serialization methods that Protobuf-Java provides? This approach serializes Protobufs into their corresponding official JSON format, as detailed in this guide: https://www.baeldung.com/java-convert-json-protobuf.

The primary reason I haven't provided a JSON deserialization method for KrotoDC data classes is that doing so would result in a slightly different JSON format compared to the original Protobufs, which seems like a poor design choice. However, adding a utility function to serialize/deserialize from the original Protobuf JSON formats could be a beneficial enhancement to this library.

**Comment by mscheong01** [comment]
I've added `toJson`, `fromJson` generated functions that provide Json conversion and released it as [1.1.0](https://github.com/mscheong01/krotoDC/releases/tag/1.1.0) 🙂

**Comment by anjanbk** [comment]
Just tried it out — there seems to be a bug in the code generator if a message has a long name 😅. Our system has a fairly deep namespace tree, and one of our messages has a somewhat large name, and it triggered the bug.

Try the following:

```
syntax = "proto3";

option java_multiple_files = true;
option java_package = "com.example.application.system.types";
option java_outer_classname = "VeryLongMessageNameProto";

package com.example.application.system.types;

message VeryVeryLongTestProtocolBuffersMessageName {
  repeated string type = 1;
}

```

The `fromJson` function applies a newline after `apply`, which doesn't compile

```
public fun KrotodcVeryVeryLongTestProtocolBuffersMessageName.Companion.fromJson(json: String):
    KrotodcVeryVeryLongTestProtocolBuffersMessageName =
    com.example.application.system.types.VeryVeryLongTestProtocolBuffersMessageName.newBuilder().apply
    { JsonFormat.parser().ignoringUnknownFields().merge(json, this@apply) }.build().toDataClass();
```

Instead, generating the following should work:

```
public fun KrotodcVeryVeryLongTestProtocolBuffersMessageName.Companion.fromJson(json: String):
    KrotodcVeryVeryLongTestProtocolBuffersMessageName =
    com.example.application.system.types.VeryVeryLongTestProtocolBuffersMessageName.newBuilder()
        .apply{ JsonFormat.parser().ignoringUnknownFields().merge(json, this@apply) }.build().toDataClass();
```

**Comment by mscheong01** [comment]
Yes, I see... Kotlin poet inserts a newline if the provided code is too long at any whitespace and for cases where the newline is inserted between `apply` and `{` the code does not compile 😢 Inserting a newline before .apply will probably fix this issue for good (I would rather not remove the whitespace between `apply` and `{` since it is convention to have it)

**Comment by mscheong01** [comment]
@anjanbk I've fixed this in #27 and released it as [1.1.1](https://github.com/mscheong01/krotoDC/releases/tag/1.1.1)

**Comment by anjanbk** [comment]
That fixed the bug, thanks.

Since Kotlin doesn't let you directly enforce the presence of a companion object's method through a generic type constraint, how would you recommend using the `fromJson` and `toJson` that you added in a generic serializer/deserializer?

**Comment by mscheong01** [comment]
That doesn't seem possible ATM 😞 
If we want to get this supported, we would have to generate conversion functions as class methods instead of extension functions.
Even then, since `fromJson` is a static function, we'll have to find another way for deserialization.

What would you rate the priority of this feature for your work?
do you have usecases where you use KrotoDC classes as other class' field types?
otherwise, you could just directly call the extension functions.

**Comment by anjanbk** [comment]
An alternative could be to not use sealed classes for oneofs, but that's a bigger change.

The general use-case we have is to define all of our data types in Protobufs, and store those types in our databases as JSONB blobs. Our Kotlin system serializes/deserializes this data when writing/reading that data in the DB. As a result, there needs to be a generic serializer that's able to handle any type as long as that type is defined correctly (e.g. `gson.fromJson(jsonString, MyType::class.java`). Gson (and most JSON serializers) require custom type adapters to handle sealed classes.

For now, we've defined a type adapter that uses reflection to handle this (something kind of like this: https://gist.github.com/krishnabhargav/7b1832eeb86aa213ba5bb239153977ea)

Would love for there to be a better way to do it though, as this is a common pattern.
